File: 0977 - Squares of a Sorted Array.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        results = []
        i, j = 0, len(nums) - 1
        while i <= j:
            if abs(nums[i]) < abs(nums[j]):
                results.append(nums[j] * nums[j])
                j -= 1
            else:
                results.append(nums[i] * nums[i])
                i += 1
        return results[::-1]

File: 0977 - Squares of a Sorted Array.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        int neg = 0;
        while (neg < nums.size() && nums[neg] < 0) {
            neg++;
        }

        vector<int> negatives;
        for (int i = neg - 1; i >= 0; i--) {
            negatives.push_back(-nums[i]); // reversed and negated to make ascending
        }

        vector<int> result;
        int i = 0, j = neg; // i for negatives, j for non-negatives in original nums

        while (i < negatives.size() && j < nums.size()) {
            if (negatives[i] < nums[j]) {
                result.push_back(negatives[i] * negatives[i]);
                i++;
            } else {
                result.push_back(nums[j] * nums[j]);
                j++;
            }
        }

        while (i < negatives.size()) {
            result.push_back(negatives[i] * negatives[i]);
            i++;
        }

        while (j < nums.size()) {
            result.push_back(nums[j] * nums[j]);
            j++;
        }

        return result;
    }
};